docs/testing.rst

```diff
@@ -0,0 +1,102 @@
+Testing Suite
+========================================================================================
+|travis| |appveyor| |coverage|
+
+.. include:: ../README.rst
+   :start-after: begin_badges
+   :end-before:  end_badges
+
+Exhale uses `pytest <https://docs.pytest.org/en/latest/>`_ for its testing suite.  To
+run the tests, Exhale uses `tox <https://tox.readthedocs.io/en/latest/>`_.
+
+.. code-block:: console
+
+   $ pip3 install tox
+   $ cd /path/to/exhale
+   $ tox
+
+.. note::
+
+   See ``tox.ini`` in the root of the repository.  You only need to install ``tox`` to
+   be able to run the tests, which will internally download the additional dependencies
+   such as ``pytest``.  The tests will be run in a virtual environment.
+
+Running Specific Tests
+----------------------------------------------------------------------------------------
+
+By default, ``tox`` will run the python tests and linting checks with
+`flake8 <http://flake8.pycqa.org/en/latest/>`_.  More formally, the default environment
+list is defined as:
+
+.. code-block:: ini
+
+   [testenv]
+   envlist = py, flake8
+
+This means that the version of python the tests are run with are the interpretor that
+you installed ``tox`` for.  To run just one or the other, use
+
+.. code-block:: console
+
+   # Run just the python tests
+   $ tox -e py
+
+   # Run just the flake8 tests
+   $ tox -e flake8
+
+There is an additional testing suite available for the documentation, which you can
+launch with ``tox -e docs``.  This will build the Exhale documentation using Sphinx with
+both the ``html`` and ``linkcheck`` builders.
+
+.. tip::
+
+   If you need to debug a test case to discern why it is failing, the vibrantly colorful
+   `ipdb <https://pypi.python.org/pypi/ipdb>`_ debugger is already installed in the
+   environment running the tests.  It is nearly identical to ``pdb``, and although it
+   is designed for IPython, it works just as well in "regular" code.  Just like with
+   ``pdb``, set a trace in the test case you are debugging:
+
+   .. code-block:: py
+
+      import ipdb
+      ipdb.set_trace()
+
+   In order to be able to achieve the break-point, you would launch the tests with
+
+   .. code-block:: console
+
+      $ tox -e py -- -s
+
+   The ``--`` signals that the command line arguments for ``tox`` are complete, and
+   everything afterward should be forwarded to the specified environment (in this case
+   to ``pytest`` since we launched ``-e py``).  The ``-s`` switch for ``pytest`` enables
+   the debugger to stop the test, without this flag the test will fail at the
+   break-point.
+
+   The same goes for other test cases, if you need to specify a flag to ``flake8`` for
+   some reason you would run ``tox -e flake8 -- --ignore XYYY`` where ``XYYY`` is the
+   linter check you want to ignore.
+
+Writing Project Test Cases
+----------------------------------------------------------------------------------------
+
+.. todo::
+
+   Finalize these docs.
+
```

Comment:
@tkhyn any chance you would be willing to fill in these docs, describing what :class:`pytest.fixture` you have enabled and what they do?

I also want to change `ExhaleTestCase::config` to be `confoverrides` for consistency, but if you're willing to write these docs maybe you could change that too?  WARNING: see hack for docs in ``testing/base.py``, there are unfortunately two class definitions.  One for the docs and one for the tests (otherwise the `ExhaleTestCase` docs were doing something really strange where it just said "alias of ExhaleTestCase" which linked to itself.

Locally I'm working on two unrelated components of the test framework

1. Validation of the hierarchy (almost finished).
2. Finalizing the CI stuff.

The docs for all the testing don't have to be perfect, I mostly just need them for myself so that after merging this I can peel off and make new project test cases as time permits.

Reply:
P.S. I am aware of the `docs` test failing right now, I know what the issue is but am awaiting a response before fixing.